### PR TITLE
[udp6] improve frame filtering (based on  #6641)

### PR DESCRIPTION
**Pull request by simonlingoogle**
Enhance frame filtering for UDP messages based on PR #6641.
- [x] Fix CI failures

### CODE REVIEW

**Comment by MarekPorwisz** [comment]
@googlebot I consent.

**Review by MarekPorwisz — APPROVED** [review]

File: src/core/net/udp6.hpp [hunk 1]
```diff
@@ -573,6 +573,18 @@ class Udp : public InstanceLocator, private NonCopyable
...
 #endif
 
+    /**
+     * This method returns whether a udp port is being used by OpenThread or any of it's optional
+     * features, e.g. CoAP API.
+     *
+     * @param[in]   aPort       The udp port
+     *
+     * @retval True when port is used by the OpenThread
+     * @retval False when the port is not used by OpenThrad.
```
Comment (jwhui) [inline src/core/net/udp6.hpp:583]:
```suggestion
     * @retval False when the port is not used by OpenThread.
```

File: include/openthread/udp.h [hunk 2]
```diff
@@ -289,6 +289,18 @@ void otUdpForwardReceive(otInstance *        aInstance,
                          const otIp6Address *aPeerAddr,
                          uint16_t            aSockPort);
 
+/**
+ * Determines if the given UDP port is exclusively opened by OpenThread API.
+ *
+ * @param[in]  aInstance            A pointer to an OpenThread instance.
+ * @param[in]  port                 UDP port number to verify.
+ *
+ * @retval true    The port is being used exclusively by OpenThread
```
Comment (jwhui) [inline include/openthread/udp.h:298]:
```suggestion
 * @retval true    The port is being used exclusively by OpenThread.
```

File: include/openthread/udp.h [hunk 3]
```diff
@@ -289,6 +289,18 @@ void otUdpForwardReceive(otInstance *        aInstance,
...
+ * @retval false   The port is not used by any of the OpenThread API or is shared (e.g. is Backbone socket ).
```
Comment (jwhui) [inline include/openthread/udp.h:299]:
```suggestion
 * @retval false   The port is not used by any of the OpenThread API or is shared (e.g. is Backbone socket).
```

**Review by librasungirl — APPROVED** [review]

**Review by bukepo — APPROVED** [review]

**Pull request merged**